nobody_home.groovy

```diff
@@ -1,164 +1,424 @@
-      log.debug("Mode is the same, not evaluating")
+    if (evt.value == "not present") {
+        handleDeparture()
+    } else {
+        handleArrival()
     }
-  }
 }
 
-def setAway() {
-  if(everyoneIsAway()) {
-    if(location.mode != newAwayMode) {
-      def message = "${app.label} changed your mode to '${newAwayMode}' because everyone left home"
-      log.info(message)
-      send(message)
-      setLocationMode(newAwayMode)
+def handleDeparture()
+{
+    log.info("${state.eventDevice} left ${location.name}")
+
+    // do nothing if someone's still home
+    if (!isEveryoneAway()) {
+        log.info("Someone is still present, no actions needed")
+        return
     }
 
-    else {
-      log.debug("Mode is the same, not evaluating")
+    // Now we set away mode. We perform the following actions even if
+    // home is already in away mode because an arrival timer may be
+    // pending, and scheduling delaySetMode() has the nice effect of
+    // canceling any previous pending timer, which is what we want to
+    // do. So we do this even if delay is 0.
+    log.info("Scheduling ${newAwayMode} mode in " + state.awayDelay + 's')
+    state.pendingOp = "away"
+    state.timerDevice = state.eventDevice
+    // we always use runIn(). This has the benefit of automatically
+    // replacing any pending arrival/away timer. if any arrival timer
+    // is active, it will be clobbered with this away timer. If any
+    // away timer is active, it will be extended with this new timeout
+    // (though normally it should not happen)
+    runIn(state.awayDelay, "delaySetMode")
+}
+
+def handleArrival()
+{
+    // someone returned home, set home/night mode after delay
+    log.info("${state.eventDevice} arrived at ${location.name}")
+
+    def numHome = isAnyoneHome()
+    if (!numHome) {
+        // no one home, do nothing for now (should NOT happen)
+        log.warn("${deviceName} arrived, but isAnyoneHome() returned false!")
+        return
     }
-  }
 
-  else {
-    log.info("Somebody returned home before we set to '${newAwayMode}'")
-  }
+    if (numHome > 1) {
+        // not the first one home, do nothing, as any action that
+        // should happen would've happened when the first sensor
+        // arrived. this is the opposite of isEveryoneAway() where we
+        // don't do anything if someone's still home.
+        log.debug("Someone is already present, no actions needed")
+        return
+    }
+
+    // check if any pending arrival timer is already active. we want
+    // the timer to trigger when the 1st person arrives, but not
+    // extended when the 2nd person arrives later. this should not
+    // happen because of the >1 check above, but just in case.
+    if (state.pendingOp == "arrive") {
+        log.debug("pending ${state.pendingOp} op already in progress, do nothing")
+        return
+    }
+
+    // now we set home/night mode
+    log.info("Scheduling ${state.modeIfHome} mode in " + state.arrivalDelay + 's')
```

Comment:
OCD: the 's' should be in double quotes to match the start of that string concatenation.